org.eclipse.winery.topologygraph/src/main/java/org/eclipse/winery/topologygraph/matching/patterndetection/ToscaComponentPatternMatcher.java

```diff
@@ -0,0 +1,148 @@
+/*******************************************************************************
+ * Copyright (c) 2021 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Apache Software License 2.0
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ *******************************************************************************/
+
+package org.eclipse.winery.topologygraph.matching.patterndetection;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import javax.xml.namespace.QName;
+
+import org.eclipse.winery.model.tosca.TEntityTemplate;
+import org.eclipse.winery.model.tosca.TEntityType;
+import org.eclipse.winery.model.tosca.TNodeTemplate;
+import org.eclipse.winery.model.tosca.extensions.OTPatternRefinementModel;
+import org.eclipse.winery.model.tosca.extensions.OTPermutationMapping;
+import org.eclipse.winery.model.tosca.extensions.OTRefinementModel;
+import org.eclipse.winery.repository.backend.NamespaceManager;
+import org.eclipse.winery.topologygraph.model.ToscaEdge;
+import org.eclipse.winery.topologygraph.model.ToscaEntity;
+import org.eclipse.winery.topologygraph.model.ToscaNode;
+
+public class ToscaComponentPatternMatcher extends ToscaPatternMatcher {
+
+    private final List<OTRefinementModel> refinementModels;
+    private final Map<OTPatternRefinementModel, List<ComponentPatternCandidate>> componentPatternCandidates;
+
+    public ToscaComponentPatternMatcher(OTPatternRefinementModel prm, NamespaceManager namespaceManager,
+                                        List<OTRefinementModel> refinementModels,
+                                        Map<OTPatternRefinementModel,
+                                            List<ComponentPatternCandidate>> componentPatternCandidates) {
+        super(prm, namespaceManager);
+        this.refinementModels = refinementModels;
+        this.componentPatternCandidates = componentPatternCandidates;
+        this.componentPatternCandidates.put(prm, new ArrayList<>());
+    }
+
+    @Override
+    public boolean isCompatible(ToscaNode left, ToscaNode right) {
+        return super.isCompatible(left, right)
+            || (!isStayingElement(left)
+            && (shareSupertype(left, right) || componentPatternsCompatible(left, right)));
```

Comment:
I would add some more parathenses and a different formatting to ease understanding this expression.